[#37] Bundle emails and link to labels

abasiri opened the pull request:
[#37] This makes an attempt at introducing bundles by linking to search for "in:inbox label:<label>"

There are still many features missing, and likely some bugs like:
* Ability to turn on and off using extension options
* Highlighting as read/unread depending on emails being bundled.
* Will entry correctly disappear if underlying emails are archived
* Potential other bugs.

![image](https://user-images.githubusercontent.com/708414/56941132-3dd64700-6ac7-11e9-8bca-33fdad88b355.png)

abasiri committed:
beginning work on bundles

abasiri committed:
Initial attempt at getting bundles working.

abasiri committed:
merge master and resolve conflicts

abasiri committed:
Additional cleanup

abasiri committed:
Make sure to only pull labels from current email list.

abasiri committed:
Should return map of labels already created.

abasiri committed:
Fix issue with links breaking for emails after first bundle.
For some reason when adding as sibling was breaking listeners for some
emails

abasiri committed:
Additional fixes to email selection when bundles are available.

russelldc commented:
To anyone wanting to test this with the default categories, as shown in @abasiri 's screenshot, you'll need to click the dropdown next to each category and allow it **In message list**:

<img width="414" alt="Screen Shot 2019-04-30 at 2 05 50 PM" src="https://user-images.githubusercontent.com/5100126/56963551-1af07500-6b51-11e9-9b8b-a418f5061ecd.png">

## Awesome start, found a few issues:

1. The title of the bundle is oddly spaced to the right. In your screenshot, this looks alright, because of your specific screen width and because you have the Important tags on. But on mine, it's odd. You can see that in the screenshot below.
2. I'm seeing the date bundle headers show up with no messages below them. This happens when an email was in your inbox from Last Month, for example, but it was bundled by label up above.
3. Shouldn't label bundles be separated by date as well? I need to do a little research, I can't remember if that's how it worked in Inbox or not.
4. There's something going **VERY** wrong with the emails that get left behind in the inbox. I'm clicking on emails in the list and it's sending me to entirely different ones.
5. Other times, clicking on either emails or the bundle does nothing at all. When I look in the console, I'm seeing errors like below, logged when I click on items in the email list:

```
m=b:452 Uncaught rWa {message: "Error in protected function: Cannot read property 'Fy' of undefined", stack: "TypeError: Cannot read property 'Fy' of undefined↵…s=AHGWq9ALJkiXFaqbqcxNHCNbW0ftvhvFvw/m=b:542:101)"}
```

Screenshot of how the bundle title offset it looks on my end:

<img width="960" alt="Screen Shot 2019-04-30 at 2 11 19 PM" src="https://user-images.githubusercontent.com/5100126/56963865-de714900-6b51-11e9-95f5-cb55028a6b38.png">

## Questions:

1. Do you plan to include the count?
2. Do you plan to show the list of senders?

Here's a screenshot of Inbox from March 2019, showing those 2 features:

![inbox](https://user-images.githubusercontent.com/5100126/56964667-905d4500-6b53-11e9-8112-911afb594e3a.png)

No need to show attachments and all that junk, I actually wasn't a fan of that in the original Inbox, anyway.

## Suggestions / Future Enhancements

1. If we're going with redirecting to the label search page rather than opening inline, then I think it would be greatly improve UX to add a back button.
2. It'd be nice to include the official colored titles as shown in that screenshot, as well as custom label colors, but those can be future tasks after this initial work.
3. Also it's not needed for this PR, but the label bundles could have a sweep button over to the right, which would solve issue #44.

russelldc commented:
Trying to figure out the 4th issue I listed above...
It seems like if there's anything bundled, if you click a non-bundled email below the bundle, you are sent to the email just below it, instead, like an indexing problem. So if you click the last email in the list, nothing happens, just an error.

abasiri commented:
I agree that there are a ton of issues with this. it seems that the way gmail handles click events is tied to email positions rather than event targets, which means when emails get bundled, clicking on emails can open up the wrong emails. I spent sometime on this to no avail.

Also the same goes for keyboard shortcuts, if you have them on, lots of errors are thrown as the emails they expect are no longer there.

abasiri committed:
temp commit

russelldc commented:
I'll try seeing if I can figure anything out, later tonight.

abasiri commented:
I think i found the culprit. They are using index of `<tr>`. Since I'm adding a `<tr>` element, it throws everything off. Converting it to a `<div>` with some styling should make it work hopefully.

russelldc commented:
By the way, the 'odd spacing' between the icon and title was just that this extra element was present,
`<td class="WA xY"></td>`, even when Important markers were turned off.

Here's one way to tell if important markers are turned on:
`if (document.querySelector('.WA.xY')​) ...`

A naive solution:
```
...
(document.querySelector('.WA.xY') ? "	<td class=\"WA xY\"></td>" : "") +
...
```

russelldc commented:
I tried switching the `<tr>` to `<div>`, but unfortunately, among other things, the `label-link` element gets lost somehow?

### Edit:

Oh! Got it, every single child element needs to be turned into a `<div>` instead of `<td>`s as well. Seems to work now, just some very minimal style adjustment needed, and doesn't get the indexing issue.

### Edit 2:

Here's the styling changes:

```css
.bundle-wrapper {
	padding-left: 20px !important;
}

.bundled-email {
	display: none !important;
}

.bundle-image {
	width: 28px;
	padding-right: 3px;
}
```

Now there's just the issue of separating those bundles by month.

### Edit 3:

Agh, of course, this looks weird right now, if you have Important marker turned on :P
Although, it's not that bad, in my opinion, if you just completely remove the `<div class="WA xY">` element. It differentiates the bundle more from a regular email.

russelldc commented:
Here's some ideas on how to create links for the label bundles by time period:

To get today's Promotions, you'd search for:
`in:inbox label:"Promotions" after:2019/4/30 before:2019/5/1`

The URL would be:
`'#search/' + encodeURIComponent('in:inbox label:"Promotions" after:2019/4/30 before:2019/5/1')`
==>
`#search/in%3Ainbox+label%3A"Promotions"+after%3A2019%2F4%2F30+before%3A2019%2F5%2F1`

jcgoble3 commented:
So I'm testing this on Firefox on Ubuntu and can't get it to work. When the Gmail page first loads, I can very briefly see label bundles, but then the page immediately (within a fraction of a second) updates and all bundles go away, and I cannot see any way to get the bundles back. Is this an Ubuntu issue or a Firefox issue? Am I doing something wrong? If I knew how to record my screen, I'd create a GIF or a short video illustrating this, but I don't know how.

Additionally, there are several display overflow issues. None of the items on the top bar, when expanded (e.g. search bar dropdown, items in the upper right), extend past the bottom of the blue bar (screenshots: [search bar](https://snag.gy/ELbOB3.jpg), [upper right](https://snag.gy/m5d1r6.jpg)), and the items on the right clobber the entire top bar, leaving it blank when closed ([screenshot](https://snag.gy/LXGPWp.jpg)) and requiring a page reload to restore it to normal. Also, the label color submenu on the labels on the left side is hidden past the edge of the left sidebar ([screenshot](https://snag.gy/4rOfuR.jpg)). Again, these may be either an Ubuntu issue and/or a Firefox issue; I am not sure what combinations you have tested on.

The glimpse I get of the bundles during page loading looks promising, so I hope we can get this to work.

russelldc commented:
@jcgoble3 The second thing you mention is a Firefox specific problem -- there was some css introduced by @boukestam that was supposed to fix some problem he saw in Firefox, but everyone who's tested it said that that css actually *causes* the problem. To fix that, just remove lines 1000-1016:

```css
/* Firefox-specific fixes */
@-moz-document url-prefix() {
	/* overflow fix emails list */
	.AO {
		overflow-y: auto;
	}

	/* overflow fix left sidebar */
	.oy8Mbf {
		overflow-y: hidden;
	}

	/* overflow fix right sidebar */
	.brC-aT5-aOt-bsf-Jw, .brC-bsf-aT5-aOt {
		overflow-y: hidden;
	}
}
```
@abasiri Could you remove that from this branch?

russelldc commented:
@jcgoble3 That's interesting, I haven't seen the bundles appear then disappear later. I doubt it would be an OS problem. I'll try this branch now in Firefox (macOS).

russelldc commented:
@jcgoble3 I'm on Firefox v66 on macOS, and the bundling is working fine for me. Could you please try my branch? It has the changes I described in comments above (except the date bundling). It should be a lot less glitchy.

https://github.com/boukestam/inbox-in-gmail/tree/ali-bundle-link-to-labels
https://github.com/boukestam/inbox-in-gmail/archive/ali-bundle-link-to-labels.zip

jcgoble3 commented:
@russelldc Just tested that. None of the CSS overflow issues are present in yours, so that's good. Bundles appear fine and show the correct emails when clicked.

However, the emails in the bundles still also show up separately as individual lines in the main inbox, which at least for me defeats the purpose. Bundles for me serve two purposes: grouping related emails together for easy reference, as well as getting those emails out of the way so the rest of my inbox can be easily used. Instead, most of my inbox is emails that are also in bundles. Ideally, an email should appear either in a bundle or in the main inbox, but not both (as in the defunct Inbox).

Compare my current actual Inbox tab ([screenshot](https://snag.gy/wkPlsM.jpg)) with your extension in Gmail ([page 1](https://snag.gy/cv06x3.jpg), [page 2](https://snag.gy/w6b0ND.jpg)). The former is easy to navigate, the latter is extremely cluttered. Getting the bundled emails out of the way is actually more important to me than the bundles themselves, because having substantially fewer lines makes the occasional critically important email easier to spot. I would suspect this could be done with some JS code that applies "display: none" CSS to each line containing a bundled label.

Speaking of which, it also seems to create a bundle for every label in the inbox (notice the Amazon bundle, which is a label I do not bundle in Inbox), which I don't want (particularly if the bundled emails are removed from the main list). There are some labels that I apply automatically via filters for purposes other than bundling, and I still want those emails to appear in the main list unbundled because they are generally important and often need immediate attention (such as Amazon). The actual Inbox allows me to choose which labels I want bundled and which ones I don't. A similar option (even if I have to dig into Firefox's add-ons settings to configure it) would be nice here.

P.S. I'm not intending to come off as overly demanding here, so apologies if I'm coming off this way. I'm just trying to offer some constructive feedback based on my workflow.

jcgoble3 commented:
On a second look, the extension does remove the last couple of days of bundled email from the main list, but older emails are still shown individually. This only lasts for a short time (a few seconds) until Gmail next updates, upon which all emails are shown in the main list.

russelldc commented:
@jcgoble3 That's so weird... they definitely stay only in their bundle, and don't show up in the list for me.

russelldc commented:
@jcgoble3 In your screenshot, I notice there's a lot of missing first letter avatars on the emails that are escaping the bundles. Could you please check the web console to see if there's any errors?

If you're not familiar, you can get to it in Firefox with `Ctrl+Shift+J`.

jcgoble3 commented:
Screenshot of the console: https://snag.gy/tVwfiN.jpg

All of the recent emails (last ~2 days) show up with the first letter avatars on initial page load. After a few seconds (simultaneously with the bundled emails reappearing), most of those avatars disappear.

russelldc commented:
@jcgoble3 I think most of that is "normal", as in it shows up without the extension.

The ones that stand out are the 3 `Empty string passed to getElementById()`. That seems relevant.

Edit: Although, I just checked, and none of the code in the extension *explicitly* calls "getElementById"

jcgoble3 commented:
I am now home and have checked both versions in Firefox for Windows 10. Behavior of both is virtually identical to Ubuntu, with the sole difference being that under @russelldc 's version, the missing avatars/icons also include all icons in the left sidebar and the snooze, archive, and star hover actions on the right side of each row (the other two hover actions show up fine).

Also, I clicked the source code link to the right of the `Empty string passed to getElementById()`, and that leads to Google's own code, which is minified. After further investigation with the console open, I was able to reliably (on Windows, not tested on Ubuntu) trigger the warning on demand by passing the mouse cursor over the buttons on the right side of the page (namely Compose, Reminder, Calendar, Tasks, and Keep). Moving the mouse cursor repeatedly from one of those to another triggered several dozen copies of the warning in under a minute.

I then deleted the extension and reloaded the page, and the Calendar, Tasks, and Keep buttons in Google's own interface (but not the native Compose button in the upper left) also consistently triggered the warning on mouseover. So it appears to be a problem in Google's code that is somehow inherited by your Compose and Reminder buttons.

russelldc commented:
@jcgoble3 missing all of the icons is likely due to my most recent commit. It was a change that loaded all images locally from the extension, instead of relying on a 3rd party image host. At one point we were using Google's original image links for those icons, then got anxious that they'd eventually stop working since Inbox was shut down, so @g60madman uploaded them to an image hosting site. Thanks a lot for pointing this out, I didn't test out that change in Firefox.

So far I've tested in Firefox macOS (before the image change), Chrome macOS and Chrome on Windows 10, and had no issues.

I'll try Firefox on Windows 10 later today... so far I'm thinking the unbundling problem you're encountering has nothing to do with browser/OS, but something to do with your specific set of emails. I'll try re-adding a bunch of old emails to my Inbox for testing.

Just in case, could you please try this in Chrome?

russelldc commented:
Just tried it out on Firefox v66 on Windows 10, and besides the missing icons issue I need to fix, the bundling is working the same as it does on Chrome for me. So I'm still thinking it has something to do with your specific collection of emails -- I'll try filling up my inbox with a bunch, old and new, to see what happens.

jcgoble3 commented:
@russelldc I tested again in Firefox/Windows 10. [This](https://snag.gy/0vbscN.jpg) is the result. Only one bundle appears (despite four labels in the inbox), and only the first "College stuff" email is hidden at first. After the first data refresh, none are hidden. I do not see any plausible explanation for this change in behavior, nor do I see anything new in the console, so this is a head scratcher for me. I also tested with all other extensions disabled with the same results.

I will try on Chrome momentarily, as soon as I can find how to load a temporary extension on Chrome.

jcgoble3 commented:
Chrome results (on Windows 10): much, much better.

On @russelldc 's branch, everything is very nicely bundled, and all images load fine. Also tested the branch of @abasiri , and it also looks great (aside from the indenting of the label names since I have important markers turned off. With abasiri's branch, bundles are dynamically added/removed as emails arrive or are cleared (there is a brief "flash" when it does so, which is annoying but a necessary evil), while russelldc's branch seems to require a page reload for this (only briefly tested, so may have missed something on that).

This behavior (of either branch) would be usable enough for me to consider closing my Inbox tab if it behaved this way in Firefox. So if we can get the behavior I'm observing in Chrome to work in my Firefox, then that would be awesome. Please let me know what I can help test; I'm a computer science student, so so I'm willing to get my hands dirty a bit (virtually speaking). I'm testing both versions (one at a time of course) via `git clone` and `git checkout`, so can easily pull new commits or tweak the code myself if you want me to try something.

Three very minor annoyances (and two bigger ones) about both versions (that wouldn't prevent me from using it) that are the side effect of using searches to implement bundling:

1. Clicking done inside a bundled email does not automatically close the email. I'm not sure if this can be fixed, since this is the normal behavior of a search in Gmail.
2. Marking the last email in a bundle as done and then backing out produces a "no search results" message instead of going back to the inbox. Also normal behavior of Gmail searches.
3. Nitpicking, but when in a bundle, could the redundant labels for "Inbox" and the label bundled be hidden with `display: none;`?
4. The bigger annoyances: There is no indication of whether a bundle contains unread emails or not, which is pretty important, particularly when new emails arrive in an existing bundle that contains emails that are being kept around for whatever reason. I think some code that checks CSS styles could detect unread emails (at least on the first page) before redrawing the page.
5. The other bigger annoyance: *All* labels are bundled, whether I want them bundled or not. I need some way to specify which labels I want bundled and which ones I don't. Some of my automatic labels are intended for a purpose other than bundling.

jcgoble3 commented:
Minor update now that a few more emails have arrived: russelldc's branch does successfully add a label bundle when new emails arrive, but does not delete them when the last email in the bundle is cleared. Also added a couple more items to the list of small annoyances above.

russelldc commented:
I just pushed up a fix for the images/icons in Firefox, on my branch `boukestam/inbox-in-gmail/tree/ali-bundle-link-to-labels`.

russelldc commented:
@jcgoble3 This is making no sense to me... there should be zero difference in bundling behavior in Firefox vs Chrome.

What versions of each browser are you using?

russelldc commented:
> Clicking done inside a bundled email does not automatically close the email. I'm not sure if this can be fixed, since this is the normal behavior of a search in Gmail.

> Marking the last email in a bundle as done and then backing out produces a "no search results" message instead of going back to the inbox. Also normal behavior of Gmail searches.

These two should be doable with event listeners via the [Gmail.js library](https://github.com/KartikTalwar/gmail.js).

Basically, for the first case, listen for the user archiving an email. If they're viewing an opened email when they do this, return them to the search page.
To solve the second case as well, also listen for when the user archives an email in the list. If the user archives the last email in the bundle, or returns to an empty search page due to the first case, return to the inbox.

> Nitpicking, but when in a bundle, could the redundant labels for "Inbox" and the label bundled be hidden with display: none;?

> The bigger annoyances: There is no indication of whether a bundle contains unread emails or not, which is pretty important, particularly when new emails arrive in an existing bundle that contains emails that are being kept around for whatever reason. I think some code that checks CSS styles could detect unread emails (at least on the first page) before redrawing the page.

Pretty sure this was intended, but no work was put towards it because the core functionality isn't working properly yet.

> The other bigger annoyance: All labels are bundled, whether I want them bundled or not. I need some way to specify which labels I want bundled and which ones I don't. Some of my automatic labels are intended for a purpose other than bundling.

This is doable, but such an edge case, I'd consider this issue complete without it, and suggest you make a PR yourself 😜

russelldc commented:
@jcgoble3 
>  Also tested the branch of abasiri , and it also looks great (aside from the indenting of the label names since I have important markers turned off. With abasiri's branch, bundles are dynamically added/removed as emails arrive or are cleared (there is a brief "flash" when it does so, which is annoying but a necessary evil), while russelldc's branch seems to require a page reload for this (only briefly tested, so may have missed something on that).  ...  This behavior (of either branch) would be usable enough for me to consider closing my Inbox tab

You're not experiencing the serious bug on abasiri's branch that I explained earlier in this thread?
The bug where clicking on any unbundled emails below a bundle will actually send you to a different email? And clicking sometimes does nothing at all, especially if you click the last unbundled email in the inbox?

Also yes, on my branch, empty bundles do not disappear without refreshing, although I'd thought the same thing happened on abasiri's branch.  I'll go double-check.

To make testing/reproduction on my side easier, could you please tell me how many emails are in your inbox right now, and how many labels?

jcgoble3 commented:
> I just pushed up a fix for the images/icons in Firefox, on my branch `boukestam/inbox-in-gmail/tree/ali-bundle-link-to-labels`.

Images in the left sidebar now work, but not the avatars on the left side of the emails (which I don't really care about).

> @jcgoble3 This is making no sense to me... there should be zero difference in bundling behavior in Firefox vs Chrome.

I don't get it either.

> What versions of each browser are you using?

Firefox 66.0.2 (which is about one version out of date, but I cannot update without losing my Inbox tab since it requires restarting the browser) and Chrome 74.0.3729.131 on Windows 10. The Ubuntu laptop is put away at the moment so I can't check for certain, but Firefox is probably the same version (since I have an open Inbox tab there as well) and I haven't tested in Chrome on Ubuntu yet.

> > The other bigger annoyance: All labels are bundled, whether I want them bundled or not. I need some way to specify which labels I want bundled and which ones I don't. Some of my automatic labels are intended for a purpose other than bundling.
> 
> This is doable, but such an edge case, I'd consider this issue complete without it, and suggest you make a PR yourself 😜

Okay then. I guess since I don't have an internship this summer, I need some kind of project to get neck-deep into. Never tried my hand at browser extensions before, but I guess it's a useful skill to learn. 😜 

> You're not experiencing the serious bug on abasiri's branch that I explained earlier in this thread?
> The bug where clicking on any unbundled emails below a bundle will actually send you to a different email? And clicking sometimes does nothing at all, especially if you click the last unbundled email in the inbox?

I didn't notice it before because I didn't have any emails below a bundle this afternoon. I unarchived an older email to test, and yes, I get a totally different email when I click it under abasiri's version. I could not reproduce the other issue in brief testing, but I may not have tested extensively enough.

> To make testing/reproduction on my side easier, could you please tell me how many emails are in your inbox right now, and how many labels?

Currently 18 emails, most in one of three labels. They are arranged in the following order, most recent first:

* 2 unlabeled (first is an email to myself, which shows up as a reminder under the extension)
* 4 with label A
* 1 with label B
* 2 with label A
* 1 with label B
* 6 with label A
* 1 with label C
* 1 with label A

abasiri commented:
Just so you don't burn extra cycles, I have a completely rewritten version
of this coming up that solves the problem with bundles

abasiri committed:
Progress on bundles. also fixes snoozed email issues

abasiri committed:
merge master and resolve conflicts

jcgoble3 commented:
Looking at these newest commits by @abasiri. The issue with an email below a bundle is fixed in Chrome. All bundles display in bold text, which I find odd.

I just noticed (because I didn't previously have such an email in my inbox while testing) that a label with a dot (`.`) in its name breaks the search link. For example, I have a label named "TheForce.net", but the corresponding bundle uses "theforce-net" rather than "theforce.net" in the search link, resulting in the bundle appearing to be empty when it isn't.

Checking on Firefox momentarily...

russelldc commented:
I'm seeing all bundles as bold (unread?) as well.

russelldc commented:
@jcgoble3 Are you still missing avatars in abasiri's latest commit? I think he's merged this one with master now, and I'd merged all the image loading changes to master. I hadn't seen that issue on my end, in Firefox or Chrome, with my bundle branch or master.

jcgoble3 commented:
> I'm seeing all bundles as bold (unread?) as well.

Yes, the last few years of using Inbox has trained me to interpret bold as unread, so it is confusing. Just in a few minutes of testing, I found myself repeatedly opening bundles thinking there were unread emails contained within.

Firefox update: abasiri's new version works much better on Firefox. I haven't fully exercised it, but the behavior of this branch now appears to be identical on both browsers. Still have no clue why it was broken in Firefox before.

Avatars are now appearing just fine in both browsers as well.

russelldc commented:
I'm still seeing the issue where emails from a different date are getting bundled into the most recent bundle. I understand that's a bit difficult to do (splitting bundles into each date), but the real problem is that the empty date label gets left behind. See "Yesterday" below:

![image](https://user-images.githubusercontent.com/5100126/57055675-c9510300-6c94-11e9-80a3-6af537dd2b63.png)

Also seeing the invisible important marker issue I mentioned before.


Good news, though, the bundles are disappearing for me without refreshing :)

abasiri commented:
@russelldc All emails bundling under the same label regardless of date is intended behaviour. Having empty date label is a bug i'll look at fixing.

russelldc commented:
@abasiri Yup, understood, I'll create a follow-up issue to address bundling by dates after this initial work is all done. Edit: maybe not, seems like this was the behavior in Inbox?

I noticed there's not an invisible important marker dom element anymore, but instead it's just a lot of left margin. Please fix that for users who don't have important markers enabled.

jcgoble3 commented:
> @russelldc All emails bundling under the same label regardless of date is intended behaviour.

Agreed, this is what the actual Inbox app did. Having a separate bundle for each date header doesn't make sense, and would result in multiple copies of the same bundle cluttering up the inbox.

abasiri committed:
Add option to enable/disable bundling.

russelldc commented:
@jcgoble3 Oh, really? That seems so off to me, I guess I just forgot. It's weird that they'd design the UX to misinform the user.

abasiri commented:
@russelldc it's not so much misinforming, as 'everything gets bundled in the location of the most recent email for that bundle'

abasiri commented:
Already, even with the bugs that exist, this is so much more productive for me than not having bundles at all.

russelldc commented:
But that **is** misinforming :P It's indicating that all of those emails are from the date shown just above the label

jcgoble3 commented:
No, it's just saying that the latest email is from that date.

If you really want that behavior (separate bundle per date), perhaps it can be an option one can enable if they wish.

abasiri commented:
I disagree. the point of bundling is to have a single entry instead of multiple to increase productivity. Say I haven't checked my emails for a few days, I come into work and check my email to find that there system was throwing hundreds of alerts over multiple days, resulting in hundreds of emails across the dates. Bundling shows me a single entry that lets me know there was a ton of alerts being thrown. I'm now informed and can dismiss them all together.

abasiri commented:
Another example, I check email after a weekend and see that people have been commenting throughout the weekend on multiple PRs that I have. I have a label for PRs. I come in on monday and see an entry for PRs. I click on it and takes me to all the emails about the PR comments, I go through them all, reply, take action etc, and then dismiss and move on to next thing on my plate.

russelldc commented:
@jcgoble3 
>No, it's just saying that the latest email is from that date.
>If you really want that behavior (separate bundle per date), perhaps it can be an option one can enable if they wish.

I'm not arguing to add this behavior, I edited that comment after you two pointed out this wasn't the original behavior in Inbox.

I'm also not arguing that this isn't the clear intention of the feature to someone who's used the product for a while... but I am arguing that this isn't actually clearly pointed out anywhere. Instead, it shows a bundle for a specific date, and the emails within aren't necessarily from that date. If that's the case, why didn't they place bundles at the top, above any date labels at all?

abasiri commented:
With latest version you have to go and manually enable it in the extension options:
![image](https://user-images.githubusercontent.com/708414/57056020-75bec500-6c54-11e9-9624-96afdfcbece5.png)

abasiri commented:
> If that's the case, why didn't they place bundles at the top, above any date labels at all?

The bundles "move" up as new email arrives. It shows that there was new activity for that bundle. But other non-bundled email might arrive later, so it will be placed above. By having the bundle be under a certain date, it says "The most recent activity in this bundle happened on this date"

PS. The 'move up' feature might be broken in my PR, haven't been able to test it yet. I think right now it requires a refresh of the page to move it up.

russelldc commented:
I noticed the default right now is Disabled, think it should default to Enabled? I notice a lot of the reviewers in the Chrome web store, Reddit, and elsewhere often say "Thank you so much, I have my Inbox back! The only thing missing is bundles..."

abasiri committed:
Default bundling to On

abasiri commented:
Sounds good to me, defaulting to on.

russelldc committed:
Locally hosted bundle image/icon, clean up quotes

abasiri committed:
Detect and fix margin for bundles when important markers are turned off.

abasiri commented:
Latest change also detects if someone has important markers and adjusts margin accordingly.

russelldc committed:
Hide empty date labels WIP (not working)

russelldc committed:
Clean up script.js, consistent spacing

abasiri committed:
Add option to extension to hide and show avatars.

abasiri commented:
while in the code to add options, I also added an option to disable showing the avatars (round circles with letters in them) I only ever found avatars useful when they had the photo of the person in them.

default is 'enabled'

![image](https://user-images.githubusercontent.com/708414/57057809-65144c00-6c60-11e9-96a3-695af9d2d896.png)

russelldc commented:
For the next few hours, I'm working on:
- Hide date/time headers if they have no visible emails beneath (because all have class `bundled-email`) - **DONE**
- URL encoding bug reported by @jcgoble3 when label contains a period, like "TheForce.net" - **DONE**
- URL encoding bug when label contains a space. Replaces with `+` instead of `%20`, because Gmail would add an additional redirect to your history making it very hard to use the back button to get back to the Inbox - **DONE**
- Hide label on emails when bundling is enabled (reported by @jcgoble3) - **DONE**
- Found a bug, labels get cut off after 23 characters with an ellipses `...` - **DONE**
- Unbundle emails, remove bundle links when bundling disabled in reloadOptions() - **DONE**
- Remove avatars when avatars are disabled in reloadOptions() - **DONE**
- Bold bundle label-link only when containing unread emails
- Show message count on bundle label like Inbox, ex: `Promotions (15)`

I'm working from this branch, rebased onto your latest:
https://github.com/boukestam/inbox-in-gmail/tree/ali-bundle-link-to-labels

Laying in bed last night, I think I may have figured out what was making me think that bundles were separated by date... does it happen in the Done folder? I have somewhat strong memories of scrolling down and seeing the bundles in Done separated by month.
@jcgoble3 may be able to answer this for certain, but I'm not sure if he's too scared to leave the 'Inbox' folder of his still-alive Inbox tab. Stay strong, we're almost there!

russelldc commented:
@jcgoble3 Not working on this right now, but I was thinking about your use case concerning labels you don't want to be bundled. Are you familiar with nested/sub-labels? Would it be destructive to your workflow to move all your unbundled labels under an 'Unbundled' parent label?

Just throwing this idea out there, by the way, nothing concrete yet... because I'm not yet sure if it's easy from our end to find out the parent label.

The benefits of this approach:
- No need to develop a complex option UI to handle this
- It would work right away, across any of your browsers/devices, without having to re-add all the labels you don't want bundled to the extension options page

Just quickly tested whether something like this would work, and came away with:

Complications of this approach:
- In Gmail, if you create a nested label, the email in the list is labelled `Parent/Child`. I'm a little disappointed there, I thought it would be just `Child`. This wouldn't be such a big deal, we could split on `/`, if it weren't for the fact that you can name labels including the character `/` ...
**Edit**: Although, actually... I guess this really isn't a problem if the extension is *only* looking for `Unbundled/`, and clipping that off, not worrying about any other parent labels.

Commit:
Hide time-row headers if only siblings are bundled and hidden

Commit:
Fix label links when containing ".", and history fix when containing " "

Commit:
Fix truncated label name in getLabels()

Commit:
Fix null has no attribute className in isEmptyDateLabel()

Commit:
Reduce font-weight of first letter avatar based on @jcgoble3 screenshot

Commit:
Hide labels when bundling is enabled

Commit:
Retain original case of label name (don't force lowercase)

Commit:
Use reloadOptions to set bundling option class for hiding labels

Commit:
Unbundle emails, remove bundle links when emailBundling disabled

Commit:
Remove avatar elements when avatars are disabled

Commit:
Improve reliability of avatar option switching, code style consistency

single quotes, space between function and ()
based on airbnb javascript style guide

abasiri commented:
@russelldc thanks for taking on some additional tasks on here. Please make sure you to use a PR to my branch to bring the changes in once you're done, so we can review them first.

russelldc commented:
@abasiri will do, probably within a few minutes.

jcgoble3 commented:
Quick replies from my phone during lunch:

I checked Inbox on my phone (using a one-version-old APK without the killswitch code), and bundles in Done still seem to appear only once in the position of the last email. Can't check on computer for a while. Also, I can safely change views (Done, snoozed, searches, labels, etc.) at will in my open Inbox tab. I just can't close the tab or reload it.

I do use nested labels occasionally, and that would work for me for an unbundled label solution.

I'll reply in more detail and test the latest commits tonight when I get home, though it will be late US Eastern Time.

russelldc commented:
@jcgoble3 Ahhh, maybe what I'm thinking of is going to Done, then having to open the most recent bundle, but scrolling down past month labels within the bundle to find what I was looking for.

It's amazing you've never accidentally refreshed your browser 😄

abasiri commented:
It will be pretty easy to have a list of labels in the options that would
get bundled, as long as you're ok just typing a list of comma separated
bundle names :)

>

russelldc commented:
Yeah, it's not a huge deal either way, just kinda annoying to have to do that again for other browsers/devices/browser profiles. I'm going to attempt the `Unbundled/` parent label feature now, so @jcgoble3 can finally close the tab, haha.

jcgoble3 commented:
I keep my email open as a pinned tab in Firefox, and Firefox itself is kept open at all times and never closed except to apply updates or reboot the computer. The tricky part has been preventing Windows 10 from automatically rebooting to run Windows updates. Had to play with "active hours" in Windows and use hibernation at night.

A comma-separated list in options would also work, but as was pointed out, a parent label would work automatically across devices as those devices gain a way to do this (I use two computers, an Android phone, and on rare occasions an Android tablet). I could see a future Inbox replacement Android app that leverages that to decide what to bundle.

Commit:
Re-arranging some functions and more code style changes

Commit:
Unbundle emails with parent label 'Unbundled'

russelldc commented:
@jcgoble3 It's ready to test over on that branch in https://github.com/abasiri/inbox-in-gmail/pull/1

Fixing these two things before I consider that feature release-ready:
- Hide labels inside bundle, but not in normal search (right now, hidden in both places) - **DONE**
- Hide display of non-Unbundle-parented labels on an Unbundled-by-parent-label email - **DONE**

Edit: Finished and pushed those two

Commit:
Hiding labels only affects Inbox and inside Bundle, hide normal labels

Commit:
Fix disappearing unbundled label

Commit:
Leave starred/pinned emails unbundled

Commit:
Bold bundle title when it contains unread messages

Commit:
Show count on bundle label, and more code re-org

Commit:
Only hide labels in Inbox and Bundle, not search pages

Commit:
Show list of senders, bold senders if unread, fix count style

Commit:
Reverse the order of senders

Commit:
Reduced draw calls significantly

Commit:
Fix sender list overflow

Commit:
De-dupe list of senders, one per unread status

abasiri committed:
Merge pull request #1 from boukestam/ali-bundle-link-to-labels

Additional work on bundling

Commit:
Override narrow width class

abasiri committed:
Merge remote-tracking branch 'inbox/ali-bundle-link-to-labels' into ali/37-bundle-link-to-labels

abasiri committed:
Fix special chars in label names when clicking on bundles.

russelldc commented:
Replying to comment left by @jcgoble3 in https://github.com/abasiri/inbox-in-gmail/pull/1

In one of my commits, I'd changed the `fixLabel` function to not just apply to all non-alphanumeric, because labels with periods in them were not working at all, and labels with spaces in them were getting users stuck in a redirect loop.

@abasiri Does the new regex work for all situations? Examples:
- `TheForce.net` (the label that was breaking with the first version, for @jcgoble3)
- `The Force.net`
- `Parent/Child`
- `Parent Label/Child Label`
- `Groceries&Shopping`
- `Groceries+Shopping`

russelldc commented:
# Updated TODO

## Fixes

1. Test that all of the above labels work with the latest commit
2. When de-duping senders, if there are multiple emails from the same person and any are unread, then show as bold, if all are read then show as normal text. (reported by @jcgoble3, strategy by @abasiri)
3. For a conversation with multiple participants, this extension always chooses the sender of the first message to display in the bundle author list. Inbox, however, displays the sender of the last message for read conversations and the sender of the first unread email for unread conversations, which is more useful. (reported by @jcgoble3)
4. Fix bundle flashing issue: insert a <style> tag and add the id of the email, then you can hide it properly. That way when gmail rewrites the <tr> tag for the emails and we lose the classes on it, it won't matter, it'll still be hidden. (strategy by @abasiri) For example:
```css
<style>
  [id=":qw"] {
    display: none;
  }
</style>
```

## Enhancements

1. Reduce the opacity of a bundle's icon image if bundle contains no unread messages. As requested by @abasiri and @jcgoble3: `I find that I want that to kind of fade to distinguish it from the unread ones better.`
2. Display unique icons (and respective font color for the label-link) for Gmail's built-in categories (Promotions, Updates, etc.) as seen in this [screenshot of Inbox](https://user-images.githubusercontent.com/5100126/57144056-1083e480-6db8-11e9-8d67-78c1ec1f062f.png). @jcgoble3 fetched the images from Inbox for us: [inbox-images.zip](https://github.com/abasiri/inbox-in-gmail/files/3140526/inbox-images.zip).
**Edit**: I just realized I had these images on my desktop all this time, since the time I'd tried to save Inbox's html/css/js before shutdown. I'd apparently just never added these particular icons to this repo. And even better news, I had grabbed the 2x res versions, @jcgoble3's zip has the lower res defaults. I'll go add my collection to the master branch, now.
3. For custom labels, use their label color for the text color of the bundle's label-link and icon (it's an image, but might be possible with some CSS hackery).

The functionality listed below would likely require [Gmail.js](https://github.com/KartikTalwar/gmail.js) for reliable observers/event listeners:

4. Display a sweep all emails icon button on bundle row on mouse hover, clicking this would archive all emails contained in bundle that aren't pinned
5. Display a back button inside the bundle (search page), to return to the inbox
6. Clicking done on an open bundled email should return you to the bundle (search page)
7. If inside an empty bundle (search page), it should return you to the inbox (including if you clicked Done on an email in the list)

russelldc commented:
Even if only fixes 1 & 4 get done right now, I'd consider it ready to merge and publish (although I just saw that issue @boukestam posted... 😞). The others could all be moved to a new issue for enhancing the bundles.

abasiri commented:
The spirit of this PR was to bundle the emails together. I think that part
is feature compete and this PR should merge. The rest of the items are
enhancements that don't really affect productivity that much.

russelldc commented:
@abasiri Broken links would definitely affect productivity :P 

I just tested the problem I described in Fix 1 (label link regex), and at least one old bug that I fixed has returned. Once you click on a label with a space in it, you can't use the back button anymore.

abasiri commented:
@russelldc ah, sorry I was accidentally looking at 1 & 4 under enhancements.

abasiri commented:
Vote for  "Inbox Reborn" with likes on this comment.

abasiri commented:
in that case we need to subs. I'll add it.

russelldc commented:
@abasiri replying in the wrong thread?

abasiri committed:
Replace spaces in label names with +

russelldc commented:
The latest commit does not fix the issue I was describing with spaces. I still can't use the back button with this label:
`Inbox In Gmail`

abasiri commented:
looking.

russelldc commented:
@abasiri This works for the label `Inbox In Gmail`, but I haven't tried it with the entire list of labels yet:

`.replace(/ /g, '+')` instead of `.replace(' ', '+')`

abasiri committed:
Fix issue with breaking back button with spaces in labels.

abasiri commented:
Looks like replacing space with "-" is the correct way. I haven't tried the other chars either. I'm looking at all the chars next.

```
label.replace(/[\/\\ ]/g, '-'))
```

russelldc commented:
Tested latest commit

Broken:
- Groceries&Shopping - partly broken, doesn't get stuck in back button loop, but adds an extra page to history that you need to back out of, when the `&` gets converted to `%26`
- Groceries & Shopping - partly broken, doesn't get stuck in back button loop, but adds an extra page to history that you need to back out of, when the `&` gets converted to `%26`
- Groceries+Shopping - broken, no results. Native gmail does this search to get it to work: `label:groceries+shopping`
- Groceries + Shopping - broken, no results. Native gmail does this search to get it to work: `label:groceries-+-shopping`
- [Test] - broken, no results. Native gmail does this search to get it to work: `label:[test]`

Works:
- Groceries-Shopping
- Groceries - Shopping
- TheForce.net
- The Force.net
- The Force .net
- The Force . net
- Parent/Child
- Parent Label/Child Label

russelldc commented:
Updated my testing list in the above comment.

Half of those that are broken get fixed by just not wrapping the label with quotation marks.

abasiri committed:
Fix special chars in labels.

abasiri commented:
Yeah between the quotes and using component encode, it fixed all the issues.

abasiri commented:
Found another issue though, if the label is too long it overflows onto the subject.

abasiri commented:
I'm gonna be away from this for the rest of the day.

russelldc commented:
@abasiri How long is the label when that happens? I'm not seeing that.

Also, what does overflow mean in this context, is the text escaping the rectangle bounds of the label?

russelldc commented:
I'll merge this now, comment later when you get a chance to explain the overflow issue, a screenshot would be great
